File: cpp/day-02.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <ranges>
#include <unordered_map>
#include <sstream>
#include <utility>

auto report_is_valid(const std::vector<int> &report) -> bool
{
    const auto n = static_cast<int>(report.size());
    bool increasing = n > 1 ? report[0] < report[1] : true;
    for (int i = 1; i < n; i++)
    {
        auto diff = report[i] - report[i - 1];
        if (increasing and 1 <= diff and diff <= 3)
            continue;
        if (not increasing and -3 <= diff and diff <= -1)
            continue;
        return false;
    }
    return true;
}

auto part_1(std::vector<std::vector<int>> &data) -> int
{
    int ans = 0;
    for (auto &report : data)
    {
        if (report_is_valid(report))
            ans++;
    }
    return ans;
}

auto part_2(std::vector<std::vector<int>> &data) -> int
{
    int ans = 0;
    for (auto &report : data)
    {
        if (report_is_valid(report))
            ans++;
        else
        {
            for (int index = 0; index < report.size(); index++)
            {
                auto v = std::vector(report);
                v.erase(v.begin() + index);
                if (report_is_valid(v))
                {
                    ans++;
                    break;
                }
            }
        }
    }
    return ans;
}

auto main(int argc, char **argv) -> int
{
    std::vector<std::vector<int>> data;
    std::string line;
    while (std::getline(std::cin, line))
    {
        std::vector<int> report;
        std::stringstream ss(line);
        int in;
        while (ss >> in)
        {
            report.push_back(in);
        }
        data.push_back(std::move(report));
    }

    std::cout << "Part 1: " << part_1(data) << std::endl;
    std::cout << "Part 2: " << part_2(data) << std::endl;
    return 0;
}

File: rust/src/bin/day-02.rs
use std::io;

fn main() {
    let data = io::stdin()
        .lines()
        .filter_map(Result::ok)
        .map(|x| {
            x.split_whitespace()
                .map(|x| x.parse::<i64>().unwrap())
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    println!("Part 1: {}", part_one(&data));
    println!("Part 2: {}", part_two(&data));
}

fn report_is_valid(report: &Vec<i64>) -> bool {
    let n = report.len();
    let increasing = if 1 < n { report[0] < report[1] } else { true };
    let mut i = 1;
    while i < n {
        let diff = report[i] - report[i - 1];
        if increasing && 1 <= diff && diff <= 3 {
            i += 1;
            continue;
        }
        if !increasing && -3 <= diff && diff <= -1 {
            i += 1;
            continue;
        }

        return false;
    }
    return true;
}

fn part_one(data: &Vec<Vec<i64>>) -> i64 {
    let mut ans = 0;
    for report in data.into_iter() {
        if report_is_valid(report) {
            ans += 1;
        }
    }
    return  ans;
}

fn part_two(data: &Vec<Vec<i64>>) -> i64 {
    let mut ans = 0;
    for report in data.into_iter() {
        if report_is_valid(report) {
            ans += 1;
        } else {
            for index in 0..report.len() {
                let mut v = report.clone();
                v.remove(index);
                if report_is_valid(&v) {
                    ans += 1;
                    break;
                }
            }
        }
    }
    return  ans;
}